Mathlib/LinearAlgebra/FiniteDimensional.lean

```diff
@@ -0,0 +1,1574 @@
+
+/-- An integral domain that is module-finite as an algebra over a field is a field. -/
+noncomputable def fieldOfFiniteDimensional (F K : Type _) [Field F] [CommRing K] [IsDomain K]
+    [Algebra F K] [FiniteDimensional F K] : Field K :=
+  { divisionRingOfFiniteDimensional F K, ‹CommRing K› with }
+#align field_of_finite_dimensional fieldOfFiniteDimensional
+
+end
+
+namespace Submodule
+
+section DivisionRing
+
+variable [DivisionRing K] [AddCommGroup V] [Module K V] {V₂ : Type v'} [AddCommGroup V₂]
+  [Module K V₂]
+
+theorem eq_top_of_finrank_eq [FiniteDimensional K V] {S : Submodule K V}
+    (h : finrank K S = finrank K V) : S = ⊤ :=
+  FiniteDimensional.eq_of_le_of_finrank_eq le_top (by simp [h, finrank_top])
+#align submodule.eq_top_of_finrank_eq Submodule.eq_top_of_finrank_eq
+
+theorem finrank_mono [FiniteDimensional K V] : Monotone fun s : Submodule K V => finrank K s :=
+  fun _ _ => finrank_le_finrank_of_le
+#align submodule.finrank_mono Submodule.finrank_mono
+
+theorem finrank_lt_finrank_of_lt {s t : Submodule K V} [FiniteDimensional K t] (hst : s < t) :
+    finrank K s < finrank K t :=
+  (comapSubtypeEquivOfLe hst.le).finrank_eq.symm.trans_lt <|
+    finrank_lt (le_top.lt_of_ne <| hst.not_le ∘ comap_subtype_eq_top.1)
+#align submodule.finrank_lt_finrank_of_lt Submodule.finrank_lt_finrank_of_lt
+
+theorem finrank_strictMono [FiniteDimensional K V] :
+    StrictMono fun s : Submodule K V => finrank K s := fun _ _ => finrank_lt_finrank_of_lt
+#align submodule.finrank_strict_mono Submodule.finrank_strictMono
+
+theorem finrank_add_eq_of_isCompl [FiniteDimensional K V] {U W : Submodule K V} (h : IsCompl U W) :
+    finrank K U + finrank K W = finrank K V := by
+  rw [← finrank_sup_add_finrank_inf_eq, h.codisjoint.eq_top, h.disjoint.eq_bot, finrank_bot,
+    add_zero]
+  exact finrank_top _ _
+#align submodule.finrank_add_eq_of_is_compl Submodule.finrank_add_eq_of_isCompl
+
+end DivisionRing
+
+end Submodule
+
+section DivisionRing
+
+variable [DivisionRing K] [AddCommGroup V] [Module K V]
+
+section Span
+
+open Submodule
+
+theorem finrank_span_singleton {v : V} (hv : v ≠ 0) : finrank K (K ∙ v) = 1 := by
+  apply le_antisymm
+  · exact finrank_span_le_card ({v} : Set V)
+  · rw [Nat.succ_le_iff, finrank_pos_iff]
+    use ⟨v, mem_span_singleton_self v⟩, 0
+    simp [hv]
+#align finrank_span_singleton finrank_span_singleton
+
+nonrec theorem Set.finrank_mono [FiniteDimensional K V] {s t : Set V} (h : s ⊆ t) :
+    s.finrank K ≤ t.finrank K :=
+  finrank_mono (span_mono h)
+#align set.finrank_mono Set.finrank_mono
+
+end Span
+
+section Basis
+
+set_option pp.explicit true in
+theorem span_eq_top_of_linearIndependent_of_card_eq_finrank {ι : Type _} [hι : Nonempty ι]
+    [Fintype ι] {b : ι → V} (lin_ind : LinearIndependent K b)
+    (card_eq : Fintype.card ι = finrank K V) : span K (Set.range b) = ⊤ := by
+  by_cases fin : FiniteDimensional K V
+  · replace fin : FiniteDimensional _ _ := fin -- porting note: fails without this line
```

Comment:
The fact this is needed is really weird...

Reply:
I've added a link to the zulip discussion.